File: leetcode/rb/2.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  dummy = ListNode.new(nil)
  cur = dummy
  tenth = 0

  while l1 && l2
    tenth, digit = (l1.val + l2.val + tenth).divmod(10)
    cur.next = ListNode.new(digit)
    cur = cur.next
    l1 = l1.next
    l2 = l2.next
  end

  rest = l1 || l2
  while rest && !tenth.zero?
    tenth, digit = (rest.val + tenth).divmod(10)
    cur.next = ListNode.new(digit)
    cur = cur.next
    rest = rest.next
  end

  cur.next = tenth.zero? ? rest : ListNode.new(tenth)

  return dummy.next
end

File: leetcode/py/2.py
class Solution:
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(None)
        cur = dummy
        carry = 0

        while l1 or l2 or carry:
            if l1:
                carry += l1.val
                l1 = l1.next
            if l2:
                carry += l2.val
                l2 = l2.next
            cur.next = ListNode(carry % 10)
            cur = cur.next
            carry //= 10

        return dummy.next




class Solution:
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(0)
        cur = dummy
        tenth = 0

        while l1 and l2:
            digit = l1.val + l2.val + tenth
            tenth = digit // 10
            digit = digit % 10
            l1 = l1.next
            l2 = l2.next
            cur.next = ListNode(digit)
            cur = cur.next

        # tenth l1 l2
        rest = l1 or l2
        while rest and tenth:
            digit = rest.val + tenth
            tenth = digit // 10
            digit = digit % 10
            rest = rest.next
            cur.next = ListNode(digit)
            cur = cur.next

        cur.next = ListNode(tenth) if tenth else rest

        return dummy.next
